control-plane-operator/hostedclusterconfigoperator/controllers/node/node.go

```diff
@@ -153,3 +163,93 @@ func (r *reconciler) nodeToNodePoolName(ctx context.Context, node *corev1.Node)
 
 	return supportutil.ParseNamespacedName(nodePoolName).Name, nil
 }
+
+func (r *reconciler) reconcileKubevirtIngress(ctx context.Context) error {
+	hcp := manifests.HostedControlPlane(r.hcpNamespace, r.hcpName)
+	if err := r.client.Get(ctx, client.ObjectKeyFromObject(hcp), hcp); err != nil {
+		return fmt.Errorf("failed to get hosted control plane %s/%s: %w", r.hcpNamespace, r.hcpName, err)
+	}
+	if hcp.Spec.Platform.Kubevirt == nil || hcp.Spec.Platform.Kubevirt.IngressPassthrowServiceSelector == nil || *hcp.Spec.Platform.Kubevirt.IngressPassthrowServiceSelector {
+		return nil
+	}
+
+	// The KubeVirt ingress services has no selector so endpoints has to be
+	// reconciled with the hosted cluster node internal IPs.
+	nodeList := &corev1.NodeList{}
+	err := r.guestClusterClient.List(ctx, nodeList)
+	if err != nil {
+		return fmt.Errorf("failed to list Nodes: %w", err)
+	}
```

Comment:
I question if this is the right spot to place this logic, and if we even need to look at the node objects directly.

It's our kubevirt cloud controller component that is setting the node's internal ip, https://github.com/kubevirt/cloud-provider-kubevirt/blob/36266069226459d57ea9fe8b0ea8959f32cd972c/pkg/provider/instances_v2.go#L142 

That IP is detected using the VMI. 

capk is doing something similar here when setting the ip on the KubeVirtMachine object, https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/1ab9fc51ae51f618cb8732245969c9e1186f9068/pkg/kubevirt/machine.go#L228 

^ that capk logic is a little bit suspect since it expects interface `[0]` to be the default interface, but it must be working or else your secondary interface tests would fail (this IP has to align with the src ip address the kubelet uses to get it's cert)

So, if we can know the node's IP address without inspecting the node directly, but instead looking at the top level capi machine objects, then maybe this logic would be a better fit in the nodepool controller. 

The nodepool controller already has access to all the machine objects [here](https://github.com/openshift/hypershift/blob/80094a10e9eab95f42c271d0233dbfa05a3142e2/hypershift-operator/controllers/nodepool/nodepool_controller.go#L713). Hopefully the machine.Status.Addresses has the info you need in it. If it doesn't, the capk might need to be updated to populate the addresses accurately.

Reply:
If we can create endpoints at the hostd cluster namespace from the nodepool controller then we can do as you say, checking Machine IP.

I will try that.